### PR TITLE
perf: Optimize CSV exporting

### PR DESCRIPTION
**Pull request by alexander-yakushev**
CSV exporting implementation has a few easy improvements to make. I managed to cut down the exporting time of @paoliniluis sample dataset (10k*400 booleans and numbers) from 8.5 sec to 5.5 sec (~1/3 reduction). Methods employed:

- More efficient iteration over multiple collections by using Java Iterators. I made custom functions `mapv` and `reduce` that keep the pleasant Clojure API while doing much better on allocations. I've put the new functions into a new `metabase.perf-utils` namespace. This is completely arbitrary; I'll gladly move them to a better place and rename the namespace.
- Improvements to number formatting. This includes caching the constructed formatter for the unique after-0 digit lengths.

In the second commit, I make some changes to `clojure.data.csv` which, let's be honest, was not written with performance in mind. First, `quote?` predicate that checks if a string should be escaped, is very sloppy for a function that is called on every single written value. Second, I modified `write-cell` because it uses `clojure.core/escape` which is quite inefficient too. This part popped up to me because my locale uses `,` as a decimal separator, so all numbers in CSV on my machine have to be escaped. This is not as important in a general case where `.` is a decimal separator.

### CODE REVIEW

**Comment by alexander-yakushev** [comment]
Found and fixed two bugs in number formatting:

1. `digits-after-decimal` worked incorrectly for negative numbers.
2. Number of digits to display was calculated incorrectly for negative numbers.

**Comment by alexander-yakushev** [comment]
Fixed another number formatting bug, this time with scientific notation. Positive integer numbers in scientific notation were always printed with no decimals, but negative numbers were printed with at most 2 decimals. E.g.:

```
12345 => 1E4
-12345 => -1.2345E4
```

After the change, integers in scientific notation will use 0, 1, or 2 decimal digits depending on how many integer digits they have.

File: src/metabase/perf_util.clj [hunk 1]
```diff
@@ -0,0 +1,71 @@
+(ns metabase.perf-util
```
Comment (crisptrutski) [inline src/metabase/perf_util.clj:1]: What do you think about renaming this something like `metabase.util.performance.iter-tools`?

There are two reasons:

1. This avoids introducing  new module (we use condo to restrict the metabase.X.* to metabase.Y.* requires)
2. This makes the nature of these two functions clearer from the namespacing where they're used.

Also, these seem like very obvious functions, is there not a library somewhere with these and more?

File: src/metabase/query_processor/streaming/csv.clj [hunk 2]
```diff
@@ -35,6 +36,38 @@
...
+  [writer data]
+  (let [separator \,
+        quote \"
+        quote? (fn [^String s]
+                 (let [n (.length s)]
+                   (loop [i 0]
+                     (if (>= i n) false
+                         (let [ch (.charAt s (unchecked-int i))]
+                           (if (or (= ch \,) ;; separator
+                                   (= ch \") ;; quote
```
Comment (crisptrutski) [inline src/metabase/query_processor/streaming/csv.clj:51]: Why are you referring to the literals here, as opposed to the bindings? Just seems more fragile if we tweak them.

**Review by crisptrutski — COMMENTED** [review]
LGTM, but I think someone with more ownership of this area should give the actual tick.

File: src/metabase/perf_util.clj (hunk 1 above)
Comment (alexander-yakushev) [reply src/metabase/perf_util.clj:1]: I'll put them wherever you tell me:). I have no preference here.

> Also, these seem like very obvious functions, is there not a library somewhere with these and more?

Perhaps there is, but I wouldn't pull a whole lib just for these two functions. Besides, this gives us the flexibility to tweak something if needed. For example, add more arities.

File: src/metabase/query_processor/streaming/csv.clj (hunk 2 above)
Comment (crisptrutski) [reply src/metabase/query_processor/streaming/csv.clj:51]: Java would implicitly make them final primitives, so I hoped as much for Clojure 🥹 

Good call to be cautious.

File: src/metabase/perf_util.clj (hunk 1 above)
Comment (crisptrutski) [reply src/metabase/perf_util.clj:1]: I'm happy even with `metabase.util.performance`, but yeah let's avoid creating a new module then.

File: src/metabase/query_processor/streaming/csv.clj (hunk 2 above)
Comment (alexander-yakushev) [reply src/metabase/query_processor/streaming/csv.clj]: I've checked; the closed over Characters are correctly downcasted to chars when comparing, so it's all OK.

File: src/metabase/util/performance.clj [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+           it4 (.iterator ^Iterable coll4)]
+       (loop [res init]
+         (if (and (.hasNext it1) (.hasNext it2) (.hasNext it3) (.hasNext it4))
+           (let [res (f res (.next it1) (.next it2) (.next it3) (.next it4))]
+             (if (reduced? res)
+               @res
+               (recur res)))
+           res))))))
+
+(defn mapv
```
Comment (bshepherdson) [inline src/metabase/util/performance.clj:62]: I'm surprised if this `mapv` is significantly better than transducer style `(into [] (map f) xs)`, for a 1-arity `f`. If we do need wider `f`s here then I think this makes sense.

**Review by bshepherdson — COMMENTED** [review]
At the top level, I'm a bit concerned that we're returning different CSVs in different locales. Hopefully that's on purpose and not an oversight.

File: src/metabase/util/performance.clj (hunk 3 above)
Comment (alexander-yakushev) [reply src/metabase/util/performance.clj:62]: Yes, this function is efficient in the multiple input collections case, the 1-collection arity exists only for completeness.

**Comment by alexander-yakushev** [comment]
Pinging @adam-james-v as I see you fixed one of the bugs recently:).

**Review by adam-james-v — APPROVED** [review]
LGTM, :)

**Pull request merged**